dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBMongoDB/AzureCosmosDBMongoDBVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,351 @@
+                        OperationName,
+                        () => this._mapper.MapFromStorageToDataModel(record, new()));
+                }
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    public Task<string> UpsertAsync(TRecord record, UpsertRecordOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(record);
+
+        const string OperationName = "ReplaceOne";
+
+        var replaceOptions = new ReplaceOptions { IsUpsert = true };
+        var storageModel = VectorStoreErrorHandler.RunModelConversion(
+            DatabaseName,
+            this.CollectionName,
+            OperationName,
+            () => this._mapper.MapFromDataToStorageModel(record));
+
+        var key = storageModel[AzureCosmosDBMongoDBConstants.MongoReservedKeyPropertyName].AsString;
+
+        return this.RunOperationAsync(OperationName, async () =>
+        {
+            var result = await this._mongoCollection
+                .ReplaceOneAsync(this.GetFilterById(key), storageModel, replaceOptions, cancellationToken)
+                .ConfigureAwait(false);
+
+            return key;
+        });
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<string> UpsertBatchAsync(
+        IEnumerable<TRecord> records,
+        UpsertRecordOptions? options = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(records);
+
+        foreach (var record in records)
+        {
+            yield return await this.UpsertAsync(record, options, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    #region private
+
+    private async Task CreateIndexAsync(string collectionName, CancellationToken cancellationToken)
+    {
+        var indexCursor = await this._mongoCollection.Indexes.ListAsync(cancellationToken).ConfigureAwait(false);
+        var indexes = indexCursor.ToList(cancellationToken).Select(index => index["name"].ToString()) ?? [];
+        var uniqueIndexes = new HashSet<string?>(indexes);
+
+        var indexArray = new BsonArray();
+
+        // Create separate index for each vector property
+        foreach (var property in this._vectorStoreRecordDefinition.Properties.OfType<VectorStoreRecordVectorProperty>())
+        {
+            // Use index name same as vector property name with underscore
+            var indexName = $"{this._storagePropertyNames[property.PropertyName]}_";
+
+            // If index already exists, proceed to the next vector property
+            if (uniqueIndexes.Contains(indexName))
+            {
+                continue;
+            }
+
+            // Otherwise, create a new index
+            var searchOptions = new BsonDocument
+            {
+                { "kind", property.IndexKind },
+                { "numLists", this._options.NumLists },
+                { "similarity", property.DistanceFunction },
+                { "dimensions", property.Dimensions }
```

Comment:
There are some common distance function types and index kinds defined in abstractions, which we should map to the values required by MogoDB here.  See e.g. `DistanceFunction.DotProductSimilarity`.
The idea is that users should be able to specify that they want `IndexKind.FLAT` and `DistanceFunction.Cosine`, and it should really work with any database that supports that, instead of having to specify a different string value for each db.  The main reason why the values here are string and not enum, is because some dbs support their own weird and wonderful index types and we need to make it extensible.

Reply:
Yes, I saw these common parameters and I was wondering if I should extend it with `IVF` algorithm, which is used in Azure CosmosDB for MongoDB. I'm also not sure how we will handle the parameters that are not common but exists in specific connector only. I think this could be a case, and in this scenario most probably users will have to use their custom `IndexKind`, passed as string, correct?

Reply:
My thoughts were that if it's a fairly common index type like FLAT or HNSW, regardless of what they are calling it, then we would map it from the type in our list.  And yeah, something like Pinecone has PGA, which is really unique and even has Pinecone in the name, so in that case the connector can have their own types as well.
Let me know if you have concerns with this approach.  I don't love it, but I feel it's a reasonable compromise between allowing customers to reuse models and config across different dbs and supporting custom types.

Reply:
That makes sense. I updated handling of index kind and distance functions to use our predefined algorithms and added mapping to Azure CosmosDB for MongoDB values.